Migrate channel related functionality to YouTube.js

absidue opened the pull request:
# Migrate channel related functionality to YouTube.js

<!-- Thanks for sending a pull request! Make sure to follow the contributing guidelines. -->
<!-- Important note, we may remove your pull request if you do not use this provided PR template correctly. -->

## Pull Request Type
<!-- Please select what type of pull request this is: [x] -->
- [x] Feature Implementation
- [x] Other - Refactor

## Related issue
closes #1359 (only local API as Invidious doesn't provide them)
closes #1357
closes #2901
closes #2851
closes #2887
closes #2805

## Description

⚠️ **This pull request is set to draft as it requires YouTube.js changes that are in the main branch but haven't been released yet. You will encounter errors if you try to use this code with YouTube.js 2.9.0** ⚠️

This pull request migrates the subscriptions and channels functionality to YouTube.js.
It adds support for channel handles, channel stats, channel tags (local API only).
It also fixes issues with fetching more videos and sorting on channels with both the local and Invidious APIs.

Channel handles are supported in two ways, as URLs `https://youtube.com/@linustechtips` or by themself `@linustechtips`, if backend fallback is enabled they also work when the Invidious API is selected.

Channel tags are shown in a list on the about tab of channels, as in NewPipe, clicking on them searches for that keyword. They are implemented as `router-link`s so opening them in a new window is possible too (middle click, ctrl + click, context menu entry).

## Screenshots <!-- If appropriate -->
Channel handle:
![handle](https://user-images.githubusercontent.com/48293849/215902358-e68f50e7-336c-4f1f-8926-ba97d2b3209f.png)

Details and tags:
![details-tags](https://user-images.githubusercontent.com/48293849/215902431-c6f707ac-32e9-4a7d-919f-116b2332d5e6.png)

Non-existent channel:
![non-existent](https://user-images.githubusercontent.com/48293849/215902458-6419a820-496e-4fc6-a555-a0e038b75e4e.png)

## Testing <!-- for code that is not small enough to be easily understandable -->
LinusTechTips URL with handle: `https://youtube.com/@linustechtips`
LinusTechTips handle: `@linustechtips`
Invalid handle (LTT with typo) `@linustehctips`
Terminated channel: `https://youtube.com/channel/UC59AcfHD5jOGqTxb-zAsahw`
Channel with playlist categories: `https://youtube.com/channel/UCez-2shYlHQY3LfILBuDYqQ `

Refresh your subscriptions, I currently haven't changed the 150 limit before switching to RSS, as I don't have anywhere near enough subscriptions to hit ratelimits (it now makes 2 requests per channel instead of 1), so if you have lots of subscriptions please try this, although I highly recommend using a VPN to test this, so that if you do get ratelimited, you can change to a different location to get away from it.

**Backup your subscriptions first**
afterwards:
Export your subscriptions as OPML
Remove all subscriptions and profiles
Import your OPML export

## Desktop
<!-- Please complete the following information-->
- **OS:** Windows
- **OS Version:** 10
- **FreeTube version:** 0.18.0

PikachuEXE commented on src/renderer/components/data-settings/data-settings.js:
```diff
@@ -1038,25 +1038,32 @@ export default defineComponent({
-          } else {
-            resolve([])
-          }
+        if (typeof channel === 'string') {
```
Need some code comment to explain :S

PikachuEXE commented on src/renderer/helpers/api/local.js:
```diff
@@ -166,6 +168,115 @@ function decipherFormats(formats, player) {
+ * Returns the channel or the channel termination reason
+ * @param {string} id
+ */
+export async function getLocalChannel(id) {
+  const innertube = await createInnertube()
+  let result
+  try {
+    result = await innertube.getChannel(id)
+  } catch (error) {
+    if (error instanceof ChannelError) {
+      result = error.message
+    } else {
+      throw error
+    }
+  }
+  return result
+}
```
For return result, maybe returning object on error case and on the caller to check if it's error is much more readable than checking the return value is string (i.e. error message returned)?

ChunkyProgrammer replied:
Agreed `return { error: error.message }` would make it more readable instead of checking if it's a string

ChunkyProgrammer commented on src/renderer/store/modules/utils.js:
```diff
@@ -288,6 +290,14 @@ const actions = {
     // If `urlType` is "invalid_url"
     // Nothing else
+
+    // Invidious doesn't support channel handles yet
```
Some instances don't yet but it is now supported as of recently https://yt.artemislena.eu/@linustechtips

ChunkyProgrammer replied:
Revisit once https://github.com/iv-org/invidious/pull/3612 is merged

ChunkyProgrammer commented:
Adding support for https://www.youtube.com/channel/UC-9-kyTW8ZkZNDHQJ6FgpwQ should be fairly easy (the header is formatted differently) and I think it'd be a good idea to support it

absidue commented:
You should now be able to view that channel as well as other special auto generated ones like:
https://youtube.com/@gaming and https://www.youtube.com/channel/UCQvWX73GQygcwXOTSf_VDVg
You'll need to build YouTube.js from source again with the latest changes on it's main branch for this to work.

ChunkyProgrammer commented:
Just wanted to add this as a potential test channel (age restricted channel for a whiskey distillery) https://www.youtube.com/channel/UCbfnHqxXs_K3kvaH-WlNlig

absidue commented:
ooh interesting so it looks like the channel page itself is age restricted but the all uploads playlist isn't (replace the first two characters in a channel URL with `UU` and use it as a playlist id)
https://www.youtube.com/playlist?list=UUbfnHqxXs_K3kvaH-WlNlig

efb4f5ff-1298-471a-8973-3d47447115dc commented:
> YouTube gaming: https://youtube.com/@gaming
Minecraft - Topic: https://www.youtube.com/channel/UCQvWX73GQygcwXOTSf_VDVg

Is this expected behavior?

https://user-images.githubusercontent.com/73130443/219893801-20028c6a-b7e4-4c6d-a28e-d27fce6df89c.mp4

absidue commented:
> >  (it now makes 2 requests per channel instead of 1)
> 
> 
> 
> Why? (Probably being dumb here)
> 
> 

yt-channel-info hard codes a bunch of stuff, allowing you to fetch the contents of a tab directly.

YouTube.js on the other hand takes a different approach, `InnerTube#getChannel()` fetches the default channel tab (for normal channels it's called home and for auto generated ones it's usually the only thing you see and is unnamed), that function returns a Channel object

`Channel#getVideos()` selects the videos tab from the list of tabs that were extracted from the first response and then calls it's endpoint

YouTube.js's approach is dynamic and also means that we could add support for all of the tabs on the Minecraft channel, without needing any upstream changes as we could do that tab selection and endpoint calling in our own code, it even exposes the Parser to us.

> Is this expected behaviour?

Yes, the auto generated gaming channel doesn't have any information to display in the About tab, no description, details, tags or related channels.

As part of the special treatment for specific channels in the future, we could use the list of popular games on that channel for the related channels list on the about page. (Yes, it's one of those channels that has an unnamed default tab)

absidue commented on src/renderer/views/Channel/Channel.js:
```diff
@@ -608,12 +848,14 @@ export default defineComponent({
-      this.channelName = this.subscriptionInfo.name
-      this.thumbnailUrl = this.subscriptionInfo.thumbnail
+      this.id = this.subscriptionInfo?.id
+      if (!hasNameAndThumbnail) {
```
Yes, that parameter is only set to true for age restricted channels, as YouTube returns the name and thumb's in the age gate response, it doesn't return that for terminated channels.

So without that parameter and if, if you aren't subscribed to the age restricted channel, setErrorMessage would overwrite the name and thumb's with null, so they wouldn't get displayed despite them being in the response returned by YouTube.

absidue replied:
Changed the variable name so it should now be a bit clearer in the code what the variable is for.

PikachuEXE commented:
<details>
<summary>**Local API**</summary>

OK:
LinusTechTips URL with handle: https://youtube.com/@linustechtips
LinusTechTips handle: @linustechtips
Invalid handle (LTT with typo) @linustehctips
Terminated channel: https://youtube.com/channel/UC59AcfHD5jOGqTxb-zAsahw
Channel with playlist categories: https://youtube.com/channel/UCez-2shYlHQY3LfILBuDYqQ

**No any** video nor playlist
YouTube gaming: https://youtube.com/@gaming
Minecraft - Topic: https://www.youtube.com/channel/UCQvWX73GQygcwXOTSf_VDVg

</details>

<details>
<summary>**Invidious API**</summary>

OK:
Terminated channel: https://youtube.com/channel/UC59AcfHD5jOGqTxb-zAsahw
Channel with playlist categories: https://youtube.com/channel/UCez-2shYlHQY3LfILBuDYqQ

OK as search result returned
LinusTechTips handle: @linustechtips
Invalid handle (LTT with typo) @linustehctips

Error returned (couldn't get channel info)
LinusTechTips URL with handle: https://youtube.com/@linustechtips
YouTube gaming: https://youtube.com/@gaming

Error returned (see screenshot)
Minecraft - Topic: https://www.youtube.com/channel/UCQvWX73GQygcwXOTSf_VDVg

![image](https://user-images.githubusercontent.com/1018543/219905367-7b1a1ccf-b1c1-4ae5-8fc9-e067869c6492.png)

</details>

absidue commented:
I'll reiterate it here despite it already being written in the pull request description and mentioned on Matrix: For auto generated channels this pull request only makes sure they can actually be displayed on the local API and extracts the available information for the header and about tabs. No videos and playlists is 100% expected for those channels, as they need special treatment, because they don't have any video or playlist tabs, this will come in a future pull request as this one is already big enough as it is. (e.g. the section headers on the live channel are links to playlists, so we could use that to populate the playlists tab and then decide which section of videos we want to display in the videos tab)

Why only that? Before this pull request both the local and Invidious APIs would error, so you would end up in a never ending fallback loop. Now at least the local API can display the header and about tabs.

absidue commented:
@PikachuEXE were you Invidious API tests done with backend fallback enabled or disabled?

absidue commented on src/renderer/views/Channel/Channel.vue:
```diff
@@ -54,13 +60,14 @@
 
           <div class="channelInfoActionsContainer">
             <ft-share-button
+              v-if="!errorMessage"
```
Changed that now, it will show it for everything except for the non-existent channel error page.

absidue commented:
@PikachuEXE You should now be able to view channels with handles on the Invidious API with backend fallback disabled or enabled. Most instances probably aren't updated yet, this one is: `https://iv.ggtyler.dev`

That means that `@linustechtips` and `https://youtube.com/@linustechtips` will both now directly open the channel regardless of what API configuration is setup.

auto generated channels like `@gaming` are still broken, Invidious doesn't support them and backend fallback isn't working correctly as the info, videos and playlists Invidious functions are all trying to fall back at the same time, I should probably change it so that the info is fetched first, that way it will fall back to the local API without ever calling the other functions if Invidious doesn't support the channel, of course if backend fallback is disabled we just have to show an error and leave it at that.

PikachuEXE commented:
I tested with backend fallback disabled
Let me re-test with backend fallback disabled again

With backend fallback enabled everything should work right???

PikachuEXE commented:
(Backend Fallback Disabled)
(Only broken/updated mentioned)

<details>
<summary>Local API</summary>

STILL **No any** video nor playlist
YouTube gaming: https://youtube.com/@gaming
Minecraft - Topic: https://www.youtube.com/channel/UCQvWX73GQygcwXOTSf_VDVg

</details>

<details>
<summary>Invidious API with https://iv.ggtyler.dev/</summary>

OK as channel page shown (and with proper error message if applicable)
LinusTechTips handle: @linustechtips
Invalid handle (LTT with typo) @linustehctips
LinusTechTips URL with handle: https://youtube.com/@linustechtips

Error returned (Missing hash key: "interactiveTabbedHeaderRenderer")
YouTube gaming: https://youtube.com/@gaming

Error returned (same error as last time)
Minecraft - Topic: https://www.youtube.com/channel/UCQvWX73GQygcwXOTSf_VDVg

</details>

absidue commented:
@PikachuEXE "no any videos nor playlist" is 100% expected, as mentioned in the pull request description, https://github.com/FreeTubeApp/FreeTube/pull/3143#issuecomment-1435926343 and in the Matrix chat.
Those channels don't have the videos or playlists tabs, so they will need special handling, which will be coming in a FUTURE pull request, to stop this one getting even larger than it already is.

PikachuEXE commented:
(Invidious API)

With fallback **disabled** now only last 2 entries have same error but with error message shown on channel view
```
Error returned (Missing hash key: "interactiveTabbedHeaderRenderer")
YouTube gaming: https://youtube.com/@gaming

Error returned (same error as last time)
Minecraft - Topic: https://www.youtube.com/channel/UCQvWX73GQygcwXOTSf_VDVg
```

With fallback **enabled** both fallback fine

Since having no video / playlist for below is fine for this PR, it can be merged
YouTube gaming: https://youtube.com/@gaming
Minecraft - Topic: https://www.youtube.com/channel/UCQvWX73GQygcwXOTSf_VDVg

PikachuEXE commented on src/renderer/helpers/api/invidious.js:
```diff
@@ -6,7 +6,7 @@ function getCurrentInstance() {
 }
 
-export function invidiousAPICall({ resource, id = '', params = {} }) {
+export function invidiousAPICall({ resource, id = '', params = {}, dontLogError = false }) {
```
```suggestion
export function invidiousAPICall({ resource, id = '', params = {}, doLogError = true }) {
```

PikachuEXE commented on src/renderer/helpers/api/invidious.js:
```diff
@@ -19,12 +19,39 @@ export function invidiousAPICall({ resource, id = '', params = {} }) {
       })
       .catch((error) => {
-        console.error('Invidious API error', requestUrl, error)
+        if (!dontLogError) {
```
```suggestion
        if (doLogError) {
```

PikachuEXE commented on src/renderer/helpers/api/invidious.js:
```diff
@@ -19,12 +19,39 @@ export function invidiousAPICall({ resource, id = '', params = {} }) {
+      params: {
+        url
+      },
+      dontLogError: true
```
```suggestion
      doLogError: false
```

PikachuEXE reviewed (changes requested):
Let's not introduce double negative code

absidue commented:
@efb4f5ff-1298-471a-8973-3d47447115dc Please grace this pull request with your approval :)

PikachuEXE commented:
> Refresh your subscriptions, I currently haven't changed the 150 limit before switching to RSS, as I don't have anywhere near enough subscriptions to hit ratelimits (it now makes 2 requests per channel instead of 1), so if you have lots of subscriptions please try this, although I highly recommend using a VPN to test this, so that if you do get ratelimited, you can change to a different location to get away from it.

Should I test this...? What are the expected results (I got VPN)

absidue commented:
Currently we have a threshold set at 150 subscriptions, after which we use RSS instead of the API to avoid ratelimiting. Ideally we shouldn't have to lower the threshold but it's only something we'll know after testing it.

If everything goes well all your subscriptions should be fetched. If it goes badly you'll get error codes returned by YouTube e.g. HTTP 429 or others

PikachuEXE commented:
I only have 123 channels subscribed, no rate limit hit with VPN enabled... with RSS
No rate limit hit without RSS too

absidue commented:
That's great to hear, looks like we don't need to adjust the limits then, yay!

ChunkyProgrammer commented:
> That's great to hear, looks like we don't need to adjust the limits then, yay!

Iirc one of the invidious team members mentioned that innertube doesnt get rate limited as often

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (commented):
1. - [x] #1359
2. - [ ] #1357 
3. - [x] #2901
4. - [ ]  #2851
5. - [ ] #2887
6. - [x] #2805 
7. - [x] Channel Handles
8. - [ ] Terminated channel
9. - [ ] Channel with playlist categories
10.  - [x] Gaming
11. - [x] Minecraft Topic
12. - [x] Age restricted channel
13. - [ ] Rate limited
14. - [x] OPML
15. - [ ] Something i noticed

Results:

1. Works as intended. Can click on a tag and searches it. Also click + enter works to open in new tab. Disable search bar disables ability to interact with tags.
2. On local API: Shows `Joined`, `Viewed` and `Location`. On Invidious API: Shows `Joined` and `Viewed`. Is this intended behavior? 
3. Works with both API's.
4. Doesn't happen anymore with both API's but i noticed that invidious fetches way more videos than local API. This wasn't the case before.
5. User reported 2 issues in this one. 3 and 6 covers this issue so i wont be testing this one.
6. Sorting works as intended on local API but not on invidious. Is this intended behavior?
7. Channel handles works with both inputs on both API's. Invalid handle shows channel does not exists. 
8. Works for both API but i found some things i would like to address. On the invidious API it shows a toast with exactly the same info that is being displayed on the channel page. This toast message isn't shown on the local API. Why is this? Also when testing this and switching between API's i sometimes first got to see `this channel doesn't exists` and after hitting enter in the search bar a few times it would show me the correct message. I think i also saw this behavior on 7 a few times but i might be wrong.
9. Works but same behavior as 6. Also with the local API i noticed an error popping up on the about page.
![duplicate-keys](https://user-images.githubusercontent.com/73130443/221388934-1fb383b7-30c6-49de-87da-0c7a029e7475.PNG)
10. Works as intended only with local API
11. Same as 10.
12. Error dump on invidious API. On local API i just shows a nice message that the contents of the channel cannot be viewed.
13. Couldnt test.
14. Works.
15. Last thing i want to address is that sometimes when u go to a channel it briefly shows the message that the channel doesn't have videos uploaded to it and after that load the videos.

absidue commented:
I've fixed the bugs in 6 and 9.

- **2.** Unfortunately Invidious doesn't return the channel location in the API response, so yes that is unfortunately expected. (we can add support for it in the future if it gets added to their API response in the future)
- **4.** I switched to the videos endpoint for Invidious so that getting more videos would work properly, as I understand it Invidious doesn't just proxy the videos list from YouTube, instead they get it from their database, so them having more I guess depends on how many they extract from their database per request.
- **8.** the delay is because it resolves the handle and channel urls when you click enter and then sends you to the channel page afterwards, so doing it multiple times was not what made it work, it just took a while. The reason why I went with this approach, is that I originally did the resolving code purely for the local API and on my machine the resolving with the local API is so fast, that it's barely noticable but now that we have it for Invidious too (which is a lot slower) also for the resolving the original URL that was passed in is needed as we need to resolve handles and urls with usernames in them `/user/{username}` and `/c/{username}` to get their channel IDs. Doing it there seemed like the most logical place because it has access to the original URL that the user entered and it also means that the channel page only ever receives real channel IDs or `@@@` for channels that don't exist.
  I should probably move that resolving to the channel page and pass the original URL to the channel page, that way you'll also see the loading symbol while it's loading instead of it looking like nothing is happening. Will be messier on the code level but should be more user friendly.
- **12.** expected as Invidious doesn't support age restricted channels (updated the pull request description now as I forgot to mention it there), we should probably make it clear to the user that the error came from Invidious and not from inside FreeTube.
- **15.** The content for the videos and playlists tabs is fetched at the same time, both of them hide the loading icon for the tabs and their content, so what is probably happening is that the playlists one finishes slightly before the videos one does, so when it hides the loading icon the content for the videos tab hasn't finished loading yet
  the fix is to implement lazy loading for the tabs, so it only loads the content when you select it for the first time but that is out of scope for this pull request

efb4f5ff-1298-471a-8973-3d47447115dc commented:
Updated the list. Did not check 8, 12 and 15.

Do you want to address 8 and 12 in this PR?
Should i create an issue for 15?

absidue commented:
Can probably do 8 inside this PR and issue for 15 might be good.

absidue commented:
@efb4f5ff-1298-471a-8973-3d47447115dc 8 should be fixed now

efb4f5ff-1298-471a-8973-3d47447115dc commented on package.json:
```diff
@@ -75,8 +75,7 @@
     "vuex": "^3.6.2",
-    "youtubei.js": "^2.9.0",
-    "yt-channel-info": "^3.2.1"
+    "youtubei.js": "^3.0.0"
```
3.1.0 has been released

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (commented):
U also might want to address this

absidue commented:
Looks like the search is broken on Invidious' side: https://github.com/iv-org/invidious/issues/3594#issuecomment-1430810670

Kazcade commented:
Might be a bit late, but I can confirm that this does indeed fix #2901 , thank you for the feature migration.
